modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2141,12 +2141,102 @@ public int compare(CodegenProperty one, CodegenProperty another) {
+                // All oneOf definitions must contain the discriminator
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema oneOf: composedSchema.getOneOf()) {
+                    if (discriminatorFound(oneOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getOneOf().size()) {
+                    return true;
+                }
+            }
+            if (composedSchema.getAnyOf() != null && composedSchema.getAnyOf().size() != 0) {
+                // All anyOf definitions must contain the discriminator because a min of one must be selected
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema anyOf: composedSchema.getAnyOf()) {
+                    if (discriminatorFound(anyOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getAnyOf().size()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Process oneOf and anyOf models in a composed schema and adds them into
+     * their parent's discriminator map if the oneOf and anyOf models contain
+     * the required discriminator. If they don't contain the required
+     * discriminator or the discriminator is the wrong type then an error is
+     * thrown
+     * @param discriminator The CodegenDiscriminator that we will add models to
+     * @param c The ComposedSchema that contains the discriminator and oneOf/anyOf schemas
+     * @param discPropName The String that is the discriminator propertyName in the schema
+     */
+    private void addComposedMappedModels(CodegenDiscriminator discriminator, ComposedSchema c, String discPropName) {
+        ArrayList<List<Schema>> listOLists = new ArrayList<List<Schema>>();
+        listOLists.add(c.getOneOf());
+        listOLists.add(c.getAnyOf());
+        for (List<Schema> schemaList: listOLists) {
+            if (schemaList == null) {
+                continue;
+            }
+            for  (Schema sc: schemaList) {
+                String ref = sc.get$ref();
+                if (ref == null) {
+                    // for schemas with no ref, it is not possible to build the discriminator map
+                    // because ref is how we get the model name
+                    // we only hit this use case for a schema with inline composed schemas, and one of those
+                    // schemas also has inline composed schemas
+                    continue;
+                }
+                Boolean discFound = discriminatorFound(sc, discPropName);
+                String modelName = ModelUtils.getSimpleRef(ref);
+                if (discFound == false) {
+                    LOGGER.error("schema {} is lacking the required string discriminator {}", modelName, discPropName);
```

Comment:
For context, this function is only run when a discriminator exists.
We are only looping through oneOf and anyOf schema lists here, so we will only see those errors if:
- a discriminator was specified
- oneOf and or anyOf schemas are defined
- one of those oneOf or anyOf schemas is missing the required discriminator

I only included the inline processing code because our code extracts inline schemas into new models. What do you want me to change?

Reply:
Ah I see. I think a comment at this point in code could go a long way. Maybe rewording the error as well?

I still have some concerns because the spec allows for moving the discriminator property out to the top level model, but that's probably an edge case that we don't need to account for immediately.

Reply:
Re: the reword recommendation...

Consider a huge spec document with 1000 $refs and heavy use of composition. If there are types such as `BrandedTool` which requires a `manufacturer` property and `HandmadeTool` which does not: the spec author may easily mistype `Hammer` where they meant `BrandedHammer` leading to a message:

> schema Hammer is lacking the required string discriminator manufacturer

Wrapping these properties in single quotes would make this stand out as model and property names. Ideally the message would read something like:

> 'BrandedTool' defines discriminator 'manufacturer', but the referenced schema 'Hammer' is missing this property.

This is probably something we'll want to add a validator for later as well.

Reply:
The error message has been changed to the suggested message